### PR TITLE
New collaborators branch

### PR DESCRIPTION
**Pull request by AbdullahAlger**
Messed up ....

### CODE REVIEW

File: app/controllers/collaborators_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  def create
+    @wiki = Wiki.find(params[:wiki_id])
+    @collaborator = @wiki.collaborators.build(collaborator_params)
+    authorize @collaborator
+
+    if @collaborator.save
+      flash[:notice] = "Collaborator added."
+    else
+      flash[:notice] = "There was an error adding the collaborator. Please try again."
```
Comment (eliotsykes) [inline app/controllers/collaborators_controller.rb]: Consider using a different flash key from `:notice` for the error message. Flash keys you can use are in the application layout template.

File: app/models/user.rb [hunk 2]
```diff
@@ -38,8 +39,10 @@ def make_wikis_public
   end
 
   def downgrade_account
-    self.update_attributes(role: "standard")
-    self.make_wikis_public
+    if self.premium?
```
Comment (eliotsykes) [inline app/models/user.rb]: Just noting out of interest that the `self.` in lines 42-44 is optional. The lines would still work identically if the `self.` was removed from each line. There's no need to remove it however. Use the style you find the clearest to follow, with and without `self.` are both good, neither are bad.

File: config/routes.rb [hunk 3]
```diff
@@ -10,7 +10,9 @@
   # You can have the root of your site routed with "root"
   # root 'welcome#index'
 
-  resources :wikis
+  resources :wikis do
+    resources :collaborators, only: [:create, :destroy]
```
Comment (eliotsykes) [inline config/routes.rb]: Good work on being specific about which actions to expose as URLs :+1:

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -39,17 +39,17 @@
...
 
     it "should allow premium users to privatize wikis" do
-      # user = User.new(role: "premium", id: 1)
-      # wiki = Wiki.new(user_id: user.id)
-      # wiki.is_owned_by?(user)
-      # expect(user.can_privatize_wiki?(wiki)).to eq true
-
+      @user.role = "premium"
+      wiki = Wiki.new(user_id: @user.id)
+      wiki.is_owned_by?(@user)
```
Comment (eliotsykes) [inline spec/models/user_spec.rb]: Line 44 can be removed, its not testing anything.

File: spec/models/user_spec.rb [hunk 5]
```diff
@@ -39,17 +39,17 @@
   context ".can_privatize_wiki?" do
```
Comment (eliotsykes) [inline spec/models/user_spec.rb]: Good work adding tests for `.can_privatize_wiki?`. Here's the code for the method:

```ruby
  def can_privatize_wiki?(wiki)
    (self.premium? && wiki.is_owned_by?(self)) || self.admin?
  end
```

Change the code to the following:

```ruby
```ruby
  def can_privatize_wiki?(wiki)
    # (self.premium? && wiki.is_owned_by?(self)) || self.admin?
    true
  end
```

Run the spec - does it still pass? If yes, can you think of test coverage you could add that's currently missing?

File: app/models/user.rb (hunk 2 above)
Comment (AbdullahAlger) [inline app/models/user.rb]: OK. I'll keep it just so I can understand it later.

File: app/controllers/collaborators_controller.rb (hunk 1 above)
Comment (AbdullahAlger) [inline app/controllers/collaborators_controller.rb]: I need to change that to `error`

File: spec/models/user_spec.rb (hunk 5 above)
Comment (AbdullahAlger) [inline spec/models/user_spec.rb]: Yeah, I will have to think of something else. This passes if it's set to true.

File: app/controllers/collaborators_controller.rb (hunk 1 above)
Comment (AbdullahAlger) [inline app/controllers/collaborators_controller.rb]: Done.

File: spec/models/user_spec.rb (hunk 5 above)
Comment (eliotsykes) [inline spec/models/user_spec.rb]: > Do you think a test for a standard? user would do it?

Good call! Yep that's the test coverage that's needed here.

Comment (eliotsykes) [inline spec/models/user_spec.rb]: ...and if you're feeling extra adventurous, a test for a premium user who does *not* own the wiki would be another good test to add.

**Pull request merged**